### PR TITLE
[ir] [refactor] Split stmt typechecking to the frontend

### LINKED ISSUE #3633: Support frontend type check for If/While statements
**Concisely describe the proposed feature**
We have the following test: https://github.com/taichi-dev/taichi/blob/21d9048d8ce03d0c8257775f33d360656dd5f949/tests/python/test_type_check.py#L30-L41, which was commented in #3441. This was because `IfExp` was expanded to regular `If` statements in our new AST builder in #3299, and we didn't have type checks for the conditions of `If/While` statements.

To make the type error messages for `If/While` statements as easy to read as normal expressions, and to uncomment the above test, we'd better support frontend type check for `If/While` statements.

### PR DESCRIPTION
**Pull request by re-xyr**
Related issue = close #3633

Remaining issues:

- `TypeError`s thrown by frontend stmt typechecker are not catched and wrapped into `TaichiCompilationError` for now

### CODE REVIEW

File: taichi/transforms/frontend_type_check.cpp [hunk 1]
```diff
@@ -0,0 +1,110 @@
+#include "taichi/ir/ir.h"
+#include "taichi/ir/frontend_ir.h"
+#include "taichi/ir/statements.h"
+
+namespace taichi::lang {
```
Comment (strongoier) [inline taichi/transforms/frontend_type_check.cpp:5]: @k-ye Shall we use this or
```cpp
namespace taichi {
namespace lang {
```

File: taichi/transforms/frontend_type_check.cpp [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+  }
+
+  void visit(FrontendSNodeOpStmt *stmt) override {
+    // Noop
+  }
+
+  void visit(FrontendAssertStmt *stmt) override {
+    if (not stmt->cond->ret_type->is_primitive(PrimitiveTypeID::i32))
+      throw TaichiTypeError(fmt::format(
+          "`assert` conditions must be of type int32; found {}. Consider using "
```
Comment (strongoier) [inline taichi/transforms/frontend_type_check.cpp:38]: The error reporting parts in `FrontendAssertStmt`, `FrontendIfStmt`, `FrontendWhileStmt` seem a bit duplicate. What about abstracting them out (a naive way might be using macros)?

**Review by strongoier — APPROVED** [review]
LGTM!

File: taichi/transforms/frontend_type_check.cpp (hunk 2 above)
Comment (k-ye) [reply taichi/transforms/frontend_type_check.cpp:38]: Hmm, I'd suggest to extract a function for this

File: taichi/transforms/frontend_type_check.cpp (hunk 1 above)
Comment (k-ye) [reply taichi/transforms/frontend_type_check.cpp:5]: SG!

File: taichi/transforms/frontend_type_check.cpp (hunk 2 above)
Comment (strongoier) [reply taichi/transforms/frontend_type_check.cpp:38]: Agreed :-)

File: taichi/transforms/frontend_type_check.cpp [hunk 3]
```diff
@@ -0,0 +1,111 @@
+#include "taichi/ir/ir.h"
+#include "taichi/ir/frontend_ir.h"
+#include "taichi/ir/statements.h"
+
+namespace taichi {
+namespace lang {
+
+class FrontendTypeCheck : public IRVisitor {
+  void report_condition_type_error(std::string stmt_name, DataType &ret_type) {
```
Comment (strongoier) [inline taichi/transforms/frontend_type_check.cpp:9]:
```suggestion
  void report_condition_type_error(std::string stmt_name, const DataType &ret_type) {
```

File: taichi/transforms/frontend_type_check.cpp [hunk 4]
```diff
@@ -0,0 +1,111 @@
...
+  void visit(FrontendAllocaStmt *stmt) override {
+    // Noop
+  }
+
+  void visit(FrontendSNodeOpStmt *stmt) override {
+    // Noop
+  }
+
+  void visit(FrontendAssertStmt *stmt) override {
+    if (not stmt->cond->ret_type->is_primitive(PrimitiveTypeID::i32))
```
Comment (strongoier) [inline taichi/transforms/frontend_type_check.cpp:44]:
```suggestion
    if (!stmt->cond->ret_type->is_primitive(PrimitiveTypeID::i32))
```

File: taichi/transforms/frontend_type_check.cpp [hunk 5]
```diff
@@ -0,0 +1,111 @@
...
+      report_condition_type_error("assert", stmt->cond->ret_type);
+  }
+
+  void visit(FrontendAssignStmt *stmt) override {
+    // No implicit cast at frontend for now
+  }
+
+  void visit(FrontendIfStmt *stmt) override {
+    // TODO: use PrimitiveType::u1 when it's supported
+    if (not stmt->condition->ret_type->is_primitive(PrimitiveTypeID::i32))
```
Comment (strongoier) [inline taichi/transforms/frontend_type_check.cpp:54]:
```suggestion
    if (!stmt->condition->ret_type->is_primitive(PrimitiveTypeID::i32))
```

File: taichi/transforms/frontend_type_check.cpp [hunk 6]
```diff
@@ -0,0 +1,111 @@
...
+  void visit(FrontendBreakStmt *stmt) override {
+    // Noop
+  }
+
+  void visit(FrontendContinueStmt *stmt) override {
+    // Noop
+  }
+
+  void visit(FrontendWhileStmt *stmt) override {
+    if (not stmt->cond->ret_type->is_primitive(PrimitiveTypeID::i32))
```
Comment (strongoier) [inline taichi/transforms/frontend_type_check.cpp:89]:
```suggestion
    if (!stmt->cond->ret_type->is_primitive(PrimitiveTypeID::i32))
```

**Review by strongoier — CHANGES_REQUESTED** [review]
1. Keep consistent usage of curly braces when there is only one statement inside a block.
2. We can further make `report_condition_type_error` into `check_condition_type`.

**Review by strongoier — APPROVED** [review]
LGTM!

**Pull request merged**